docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,473 @@
+
+## Current implementations and challenges with the Rust components
+
+* Some apps only care about a subset of the engines - lockbox is one such app
+  and only cares about a single collection/engine. It might be the case that
+  lockbox uses a generic application-services library with many engines
+  available, even though it only wants logins. Thus, the embedding application
+  is the only thing which knows which engines should be considered to "exist".
+  It may be that the app layer passes an engine to the sync manager, or the
+  sync manager knows via some magic how to obtain these handles.
+
+* Some apps will use a combination of Rust components and "legacy"
+  engines. For example, iOS is moving some of the engines to using Rust
+  components, while other engines will be ported after delivery of the
+  sync manager, if they are ported at all. We also plan
+  to introduce some rust engines into desktop without integrating the
+  "sync manager"
+
+* The rust components themselves are designed to be consumed as individual
+  components - the "logins" component doesn't know anything about the
+  "bookmarks" component.
+
+There are a couple of gotchyas in the current implementations too - there's an
+issue when certain engines don't yet appear in meta/global - see bug 1479929
+for all the details.
+
+The tl;dr of the above is that each rust component should be capable of
+working with different sync managers. That said though, let's not over-engineer
+this and pretend we can design a single, canonical thing that will not need
+changing as we consider desktop and iOS.
+
+## State, state and more state. And then some state.
+
+There's loads of state here. The app itself has some state. The high-level
+Sync Manager component will have state, the low-level component will have state,
+and each engine has state. Some of this state will need to be persisted (either
+on the device or on the storage servers) and some of this state can be considered
+ephemeral and lives only as long as the app.
+
+A key challenge will be defining this state in a coherent way with clear
+boundaries between them, in an effort to allow decoupling of the various bits
+so Desktop and iOS can fit into this world.
+
+This state management should also provide the correct degree of isolation for
+the various components. For example, each engine should only know about state
+which directly impacts how it works. For example, the keys used to encrypt
+a collection should only be exposed to that specific engine, and there's no
+need for one engine to know what info/collections returns for other engines,
+nor whether the device is currently connected to WiFi.
+
+A thorn here is for persisted state - it would be ideal if the low-level
+component could avoid needing to persist any state, so it can avoid any
+kind of storage abstraction. A solution here might be that the low-level
+component can delegate state storage to the high-level component in an
+opaque way.
```

Comment:
I will add that this is counter to the way many of our other components act, which is to handle their own storage transparently.  FxA is an outlier in this regard, but maybe we can follow some of its patterns around persisting stage?